src/Tools/dotnet-monitor/Egress/EgressFactory.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```

Comment:
The reason why this code is in dotnet-monitor and not in the REST server is that it is an implementation detail of dotnet-monitor only (especially the configuration schema, how the egress options are consumed from configuration, and how the options converted to the EgressProvider classes from the monitoring assembly). My understanding is that the REST server will eventually be used as a component that other projects can incorporate so that they can expose the monitoring capabilities via the HTTP API. No other project is going to want to wholesale take on the configuration schema and implementations details for egress established for dotnet-monitor.

Reply:
oh whoops, I put this comment in the wrong place, sorry about that! I meant to say can we move all the Egress code in Microsoft.Diagnostics.Monitoring to the REST server. I had no qualms about the dotnet-monitor/REST server split : )

Reply:
If they are not going to be part of Microsoft.Diagnostics.Monitoring, I feel that it makes more sense for them to be in dotnet-monitor and then just fold all of the ConfiguredEgressProvider (from dotnet-monitor) and EgressProvider (currently from Microsoft.Diangostics.Monitoring) together into one implementation rather than having ConfiguredEgressProvider be a wrapper around each EgressProvider. The REST server really doesn't care about those implementations and it shouldn't force consumers of the REST server to pull in Azure storage assemblies unnecessarily.

I'll think about this a bit more before I go make the change. Another thing that I'm keeping in mind is that triggers (at the level of dotnet-monitor) will need to understand how to egress collected artifacts. I may be able to decouple egress from triggers in the Microsoft.Diagnostics.Monitoring assembly by just having the trigger pipeline report "this trigger in this state of the pipeline was satisfied" and then correlate that to the egress description in the dotnet-monitor configuration. 

@wiktork, do you have any thoughts on the me merging all of the egress notions into dotnet-monitor?

Reply:
My initial thinking here was that we would make egress an Api that we could reuse in other contexts, similar to the way the Pipeline Api works. The benefit here is that different consumers (especially those of the trigger pipeline) wouldn't have to rewrite egressing. We can keep trigger actions decoupled from egress either way.

Assuming we don't want Egress to be an API but rather an implementation detail, I think dotnet-monitor would be the appropriate place for these. I think keeping the Azure related egress in a separate assembly is still ideal though.

Reply:
> I feel that it makes more sense for them to be in dotnet-monitor

I am fine with either for now and I assume it wouldn't be too hard to move it in the future if needed. My rough guess is that REST server would make sense for functionality you would expect to work if the collection logic was loaded as a library inside a customer app (future scenario) vs. dotnet-monitor is a way to host the collection logic as a stand-alone 2nd process.

>  I may be able to decouple egress from triggers

I hope and anticipate we should be able to keep them decoupled. Its not apparent to me that a trigger needs to understand where the data is going. It only needs to determine when the data should flow or when it should stop flowing.

Reply:
> Assuming we don't want Egress to be an API but rather an implementation detail

My rule of thumb is that we would pay the cost to design sharable abstractions when we can identify scenario beyond dotnet-monitor that will make use of it. I don't expect any of the dotnet-trace,counters,dump, etc tools to have extensible egress so I don't see that 2nd scenario yet. Did you have one in mind?

Reply:
I've moved all egress to the dotnet-monitor project.